Feature: Rec model twitter enhancement

davemfish opened the pull request:
Fixes #1551 

Much of the work in this PR relates to constructing the twitter quadtree, which is something that only needs to happen once. Many of the scripts in `invest/scripts/recreation_server` pertain to that.

Some optimizations were made to speedup quadtree construction, to save memory and to utilize multiple cores. This was necessary in order to build the twitter quadtree on Sherlock. It also benefits users because "local" quadtrees are constructed during every model run.
* `buffered_numpy_disk_map.py`: appending to numpy arrays on disk without loading them entirely into memory
* `out_of_core_quadtree.pyx`: using multiple processes to flush data to disk during quadtree construction.

In `recmodel_server.py` a new `RecManager` class manages the connection to clients and delegates to userday calculations to multiple instances of `RecModel` servers, one for twitter and one for flickr. 

`recmodel_client.py` updates are mostly focused on the additional output data (TUD_results.gpgk) and a slightly modified response variable in the least-squares regression model. 

These misc issues are also addressed:
[Rec model logging related to size of server request #1285](https://github.com/natcap/invest/issues/1285)
* Log messages from the server are queued and retrieved by the client during the userday calculations so users can see progress updates.
* This includes the overall size of the query represented by the number of points found in the AOI bounding box.

[Rec server crashes when out of memory #1284](https://github.com/natcap/invest/issues/1284)
* Query sizes are limited now by a parameter set in the server startup script. This prevents people from making unreasonably large queries.
* userday calculations happen in child processes so that even if one of them does crash, the server process itself does not crash.

[Rec model text output should be tabular #1118](https://github.com/natcap/invest/issues/1118)
* Done

[Rec model server health checks #1429](https://github.com/natcap/invest/issues/1429)
* I added a test for this in our test suite. We may want to improve on that in the future so that we can find out if the server is down at other times, but for now, we'll notice it during GHA runs.

[Visitation: keep all fields in user-supplied AOI #1366](https://github.com/natcap/invest/issues/1366)
* Done

test-data PR: https://bitbucket.org/natcap/invest-test-data/pull-requests/32
sample-data PR: https://bitbucket.org/natcap/invest-sample-data/pull-requests/37
UG PR: https://github.com/natcap/invest.users-guide/pull/164

## Checklist
- [x] Updated HISTORY.rst and link to any relevant issue (if these changes are user-facing)
- [x] Updated the user's guide (if needed)
- [x] Tested the Workbench UI (if relevant)

emlys commented on src/natcap/invest/recreation/recmodel_server.py:
```diff
@@ -540,59 +827,96 @@ def _file_len(file_path):
+    LOGGER.info('%d lines', total_lines)
+
+    # On a single CPU, flushing to disk is the main bottleneck.
+    # For large trees (i.e. twitter) more than 80% of wall-time
```
What is meant by wall-time?

davemfish replied:
I just meant "real" time elapsed, like watching the clock on the wall 🕐 . As opposed to time-elapsed * n_CPUs. I guess it's redundant since I said I tested on a single CPU anyway.

emlys commented on src/natcap/invest/recreation/recmodel_server.py:
```diff
@@ -540,59 +827,96 @@ def _file_len(file_path):
+    if n_flush_processes < 1:
+        n_flush_processes = 1
+
+    ooc_qt = out_of_core_quadtree.OutOfCoreQuadTree(
```
Maybe a silly question, but what is the actual meaning of "out of core" for a quad tree?

davemfish replied:
I remember asking Rich the same thing years ago. "out of core _memory_". I guess its a term for algorithms that use data too large for memory and are designed to use the disk as an alternative. When I talk about it I sometimes call it a "filesystem-based" quadtree.

emlys reviewed (changes requested):
This is awesome @davemfish! I had a bunch of questions and some minor suggestions. I might have more questions later, but I'm going to cut off this huge review here and mull it over more.

claire-simpson commented on src/natcap/invest/recreation/out_of_core_quadtree.pyx:
```diff
@@ -345,6 +348,25 @@ class OutOfCoreQuadTree(object):
+        else:
+            return sum([
+                self.nodes[index].estimate_points_in_bounding_box(bounding_box)
+                for index in xrange(4)])
```
I believe `xrange` was removed in python 3 and replaced by `range` (looks like `xrange` is used in a bunch of these functions)

davemfish replied:
Hmm that's interesting! It looks like `xrange` is still available in `Cython` for some reason. I wonder if it is any different than Python 3's `range`. I feel like it must be okay to update these to `range`.
https://github.com/cython/cython/blob/master/CHANGES.rst#024-2016-04-04

claire-simpson commented on src/natcap/invest/recreation/recmodel_client.py:
```diff
@@ -607,72 +670,126 @@ def _retrieve_photo_user_days(
+    LOGGER.info('Please wait for server to calculate PUD and TUD...')
+    client_id = uuid.uuid4()
+
+    def wrap_calculate_userdays():
```
This is pretty minor, but this function could be edited to:

def wrap_calculate_userdays():
&nbsp;&nbsp;&nbsp;with Pyro5.api.Proxy(server_url) as proxy:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return proxy.calculate_userdays(zip_file_binary, start_year, end_year, list(datasets), client_id)

[According to Pyro5](https://pyro5.readthedocs.io/en/latest/clientcode.html#proxies-connections-threads-and-cleaning-up), this is the preferred method of using pyro proxies.

davemfish replied:
Thanks, looks like a great use-case for a context manager

claire-simpson commented on src/natcap/invest/recreation/recmodel_client.py:
```diff
@@ -1264,48 +1364,132 @@ def _ogr_to_geometry_list(vector_path):
+            'Cannot compute regression because either PUD_YR_AVG'
+            'or TUD_YR_AVG are all 0s')
+
+    for feature in target_layer:
```
This is maybe a silly question, but are we confident that equivalent FIDs of the PUD and TUD vectors are associated with the same geometry?

davemfish replied:
After working on this for a while, I'm not so sure what to do. There are a couple issues that arise in constructing our own unique ID field, like a "polygon id".
1. In this `for feature in target_layer:` block, if we're not relying on the features being in the same order (i.e. matching FIDs), then for each feature, we need to iterate over all the other features in the two other layers to find the one with a matching "polygon ID".
2. When we use `zonal_stats` in other parts of this model, it returns results keyed by FID, so we're still using FID to join those stats dictionaries back to the geometries.

That said, it still might be useful to make a "polygon_id" field in all the output vectors so that users can do table joins if they wish. But internally in the model, I think it might be okay to use FIDs.

claire-simpson replied:
Ok great, that all makes sense to me, sticking with FIDs seems fine!

claire-simpson commented on src/natcap/invest/recreation/recmodel_client.py:
```diff
@@ -1364,47 +1553,41 @@ def _build_regression(
+    # Logit transformation.
+    # The smallest non-zero value is added to the array because
+    # the array is likely to include zeros.
+    epsilon = response_array[response_array > 0].min()
```
Making a note of what I asked earlier: why set `epsilon` to the min positive value of `response_array`, rather than some arbitrary small value? Also, should there be a check here to ensure that array_adj is not 1 (which would create a 0 denominator)? If `array_adj` == 1, could add a small term

claire-simpson commented on tests/test_recreation.py:
```diff
@@ -771,134 +1001,118 @@ def test_overlapping_features_in_polygon_predictor(self):
+        # results['r_sq_adj'] = r_sq_adj
+        # results['std_err'] = std_err
+        # results['dof'] = dof
+        # results['se_est'] = se_est
```
Curious if there is a reason not to test these?

davemfish replied:
No good reason. I added assertions for these.

claire-simpson commented on src/natcap/invest/recreation/buffered_numpy_disk_map.py:
```diff
@@ -36,10 +56,13 @@ def __init__(self, manager_filename, max_bytes_to_buffer):
                 binary data as needed.
             max_bytes_to_buffer (int): number of bytes to hold in memory at
                 one time.
+            n_workers (int): if great than 1, number of child processes to
```
```suggestion
            n_workers (int): if greater than 1, number of child processes to
```

claire-simpson commented on src/natcap/invest/recreation/out_of_core_quadtree.pyx:
```diff
@@ -51,6 +51,8 @@ class OutOfCoreQuadTree(object):
                 to store the node data across the entire quadtree
             pickle_filename (string): name of file on disk which to pickle the
                 tree to during a flush
+            n_workers (int): if great than 1, number of child processes to
```
```suggestion
            n_workers (int): if greater than 1, number of child processes to
```

claire-simpson commented on src/natcap/invest/recreation/recmodel_server.py:
```diff
@@ -254,28 +481,36 @@ def _calc_aggregated_points_in_aoi(
         aoi_extent = aoi_layer.GetExtent()
         aoi_ref = aoi_layer.GetSpatialRef()
 
+        # TODO: replace all this with pygeoprocessing.transform_bounding_box
```
Still a TODO?

davemfish replied:
Thanks, I replaced this with `pygeoprocessing.transform_bounding_box`

claire-simpson commented on src/natcap/invest/recreation/recmodel_server.py:
```diff
@@ -97,47 +94,241 @@ def try_except_wrapper(*args, **kwargs):
+    def fetch_aoi_workspaces(self, workspace_id, server_id):
+        """Download the AOI in the workspace specified by workspace_id.
+
+        Cosntructs the path using the server's self.local_cache_workspace.
```
```suggestion
        Constructs the path using the server's self.local_cache_workspace.
```

claire-simpson commented on src/natcap/invest/recreation/recmodel_server.py:
```diff
@@ -148,104 +339,140 @@ def get_valid_year_range(self):
+                where coordinates are WGS84 decimal degrees.
+
+        Returns:
+            int
```
Could use a description

claire-simpson reviewed (changes requested):
Very cool @davemfish!

davemfish commented on Makefile:
```diff
@@ -2,15 +2,15 @@
+GIT_TEST_DATA_REPO_REV      := 26e7bb702e6218ae0b2d553f7402ff2c8e1fccc4
 
-GIT_UG_REPO                 := https://github.com/natcap/invest.users-guide
+GIT_UG_REPO                 := https://github.com/davemfish/invest.users-guide
```
Hmm, probably better to merge the UG PR and then update this again here. https://github.com/natcap/invest.users-guide/pull/164

davemfish commented on scripts/recreation_server/build_twitter_quadtree.sh:
```diff
@@ -0,0 +1,35 @@
+git checkout exp/REC-twitter
+
+set -x  # Be eXplicit about what's happening.
+FAILED=0
```
I think it's an unused variable, I'm removing it.

davemfish commented on scripts/recreation_server/build_twitter_quadtree.sh:
```diff
@@ -0,0 +1,35 @@
+# ----------------Commands------------------------
+
+
+CONTAINER=ghcr.io/davemfish/invest:exp.rec-twitter
```
Yeah, perhaps. Though maybe it's okay for the next person who needs to use the script to update it to whatever version of invest they want to use? It might be `3.15.0` but it might also be another bugfix/feature branch somewhere.

emlys replied:
Okay, in that case it might help to add a comment about that, reminding future us to check/update this version?

davemfish replied:
Added a comment.

davemfish commented on scripts/recreation_server/copy_quadtree_to_gcs.sh:
```diff
@@ -0,0 +1,15 @@
+#SBATCH --mail-type=ALL
+#SBATCH --mail-user=davefisher@stanford.edu
+#SBATCH --partition=hns,normal
+#SBATCH --output=/scratch/users/dfisher5/slurm-logfiles/slurm-%j.%x.out
```
I removed my email and filepath.

davemfish commented on scripts/recreation_server/launch_recserver.sh:
```diff
@@ -1,4 +1,6 @@
-conda activate ./invest/env/
-nohup python -u execute_recmodel_server.py > nohup_recmodel_server.txt 2>&1 &
+# source /home/davemfish/miniconda3/etc/profile.d/conda.sh
+# mamba activate /usr/local/recreation-server/invest_3_15_0/invest/env
```
I updated the comment, 
```
# activate a python environment before calling this script:
# mamba activate /usr/local/recreation-server/invest_3_15_0/invest/env
# I think I had a hard time doing the activation within this script, for some reason.
```

davemfish commented on scripts/recreation_server/setup_vm.sh:
```diff
@@ -0,0 +1,35 @@
+cd /usr/local/recreation-server
+mkdir invest_3_15_0
+cd  invest_3_15_0
+git clone https://github.com/davemfish/invest.git
```
Yeah, I think that makes sense. And then I think we should remember to update the server to `3.15.0` when it's released.

davemfish commented on scripts/recreation_server/setup_vm.sh:
```diff
@@ -0,0 +1,35 @@
+git checkout feature/REC-twitter-client
+mamba create -p ./env python=3.12
+mamba activate ./env
+mamba install gdal pygeoprocessing numpy
```
I'll add some restrictions based on requirements.txt.

davemfish commented on scripts/recreation_server/setup_vm.sh:
```diff
@@ -0,0 +1,35 @@
+
+# After setting up a new VM or new server cache, initialize the cron
+# job that periodically clears out the cache:
+# invest/scripts/recreation_server/cron_find_rm_cached_workspaces.sh
```
Yeah it probably could. This `setup_vm.sh` script was meant to be more of a record of how we set it up, not necessarily something someone would just execute. The cron setup I did interactively by editing the `crontab` file, but there's probably a programmatic way to do it too. I'll add a call to launch the python process, for completeness.

davemfish commented on src/natcap/invest/recreation/_utils.py:
```diff
@@ -0,0 +1,30 @@
+    """
+    with BytesIO() as file_stream:
+        numpy.save(file_stream, numpy_array, allow_pickle=False)
+        return file_stream.getvalue()
```
I played around with this a bit. I think the extra header info we get from `numpy.save` includes the datatype of the array. That's convenient, because with `tobytes`, and no header, we need to pass the `dtype` to the subsequent `numpy.frombuffer`. Not a big deal because we're defining a custome `dtype` for these arrays already, but it is a bit more passing around variables, and updating some tests that don't strictly use arrays of the same type as our "real" ones.

I'm inclined to leave the functions as-is, but I'm glad to understand this a bit better. What do you think?

emlys replied:
That makes sense, thanks for investigating! I guess I was expecting there would be a way to directly export a complete array to bytes without needing the file stream stuff, but it seems like there isn't. https://numpy.org/doc/2.2/reference/routines.io.html

davemfish commented on src/natcap/invest/recreation/recmodel_client.py:
```diff
@@ -234,8 +232,44 @@
         },
-        "monthly_table.csv": {
-            "about": gettext("Table of monthly photo-user-days."),
+        "TUD_results.gpkg": {
```
That's what we settled on in the design doc. I won't be surprised if users request the option to choose. And we can consider adding that feature.

davemfish commented on src/natcap/invest/recreation/recmodel_client.py:
```diff
@@ -607,72 +670,126 @@ def _retrieve_photo_user_days(
+            raise ValueError(
+                f"Start year must be between {min_year} and {max_year}.\n"
+                f" User input: ({start_year})")
+        if not min_year <= int(end_year) <= max_year:
```
The constraints in the MODEL_SPEC should be good for `validate`. So this is an extra precaution if `validate` is never called, like for a Python user.

emlys replied:
This is fresh in my mind because it just came up in Megan's PR too - a while back we intentionally deduplicated validation in execute vs. validate (https://github.com/natcap/invest/pull/1375), so I'm hesitant to reintroduce that. 

Whether or not this check is here, can the server also check on its side that the years requested are valid? I think that would be good to have either way. And maybe a 400 error would be helpful to Python users.

davemfish replied:
> This is fresh in my mind because it just came up in Megan's PR too - a while back we intentionally deduplicated validation in execute vs. validate (#1375), so I'm hesitant to reintroduce that.

This check is actually not new in this PR, it may have just moved around a bit. But anyway, I think it's specifically not in `validate` because it involves a call to the server, which I think is probably not a good idea to do within `validate`.

> Whether or not this check is here, can the server also check on its side that the years requested are valid? I think that would be good to have either way. And maybe a 400 error would be helpful to Python users.

Yes, I think it's a good suggestion to do some validation on the server. In theory someone could use the `recmodel_server` API exclusively to do local queries, and in those cases, validating their input dates would be nice. I'm adding that validation in `RecModel._calc_aggregated_points_in_aoi`, which is the most likely entry-point for that use-case.

I still think it's useful to keep this check in `recmodel_client` -- where the client requests the valid date range from the server -- because for users of `recmodel_client`, I think that's a better experience than having an exception get raised on the server and having to pass it back to the client. If invalid dates are requested, we can avoid sending the AOI to the server, starting subprocesses, etc.

But...that whole idea made me question what happens if an exception occurs on the server in one of the worker processes, and I realized we weren't handling it gracefully. So I've added to `RecManager.calculate_userdays` to make sure the process running the Pyro daemon doesn't re-raise an exception, and passes it back to the client to raise instead. I could not determine a way to test that properly. Mocking a server function to raise an error wouldn't work because all the server objects in our tests get created in separate processes, so I don't think mocks can reach them. I tested the code manually but did not write a new test for it. @emlys

davemfish commented on tests/test_recreation.py:
```diff
@@ -38,6 +42,10 @@
 LOGGER = logging.getLogger('test_recreation')
 
+# These are defined in the model spec for validating start and end year
+MIN_YEAR = 2012
```
Yes, good idea!

davemfish commented on src/natcap/invest/recreation/buffered_numpy_disk_map.py:
```diff
@@ -5,17 +5,37 @@
 
 
+def _npy_append(filepath, array):
+    """Append to a numpy array on disk without reading the entire array."""
```
Yeah, this was fun to figure out! It was a process of experimenting with quadtree builds and observing where the most time was spent. I did that with logging and analyzing the logs to generate some plots. It became clear that flushing to disk was the most time-consuming part, and that flush times grew as the quadtree grew larger. Take a look at this doc, I think it illustrates things pretty well: https://docs.google.com/document/d/1focUpKf62UbrlEsjfVVg3fP8no8ltaVbOAGdwrIII5I/edit?tab=t.0#heading=h.ihf4zg5j0lkb

davemfish commented on tests/test_recreation.py:
```diff
@@ -1280,6 +1408,25 @@ def test_validate_predictor_types_incorrect(self):
+class RecreationProductionServerHealth(unittest.TestCase):
+    """Health check for the production server."""
+
+    def test_production_server(self):
```
And this is now the *only* test that pings a remote server!

davemfish commented on src/natcap/invest/recreation/recmodel_server.py:
```diff
@@ -97,47 +94,241 @@ def try_except_wrapper(*args, **kwargs):
+                    ooc_qt_picklefilename, cache_workspace)
+        else:
+            LOGGER.info(
+                f'{ooc_qt_picklefilename} not found, constructing quadtree')
```
I don't think it can be `None` at this point because both the csv and pickle cannot be `None`. And if a csv is given, then it's hashed and `ooc_qt_picklefilename` is constructed using the hash. So the pickle path will be defined, but may not exist yet.

Honestly I don't like all the conditionals involved here, but that's where I ended up. It's a product of wanting to preserve how the original `RecModel.__init__` was designed for the flickr case:
* to take a CSV input and construct a quadtree if it doesn't already exist, or if the CSV hash is different from the hash used to create the quadtree last time. Or to skip quadtree construction when the hash matches (e.g. when we're just restarting the server process).

Combined with the new need to design for the twitter case:
* a quadtree already exists and we pass in the name of the pickle containing the quadtree index.

emlys replied:
Ah okay, I understand now. I think this log message might be confusing to me if I was not super familiar with what's happening here - may I suggest a longer message like 
```suggestion
                f'Pickle file {ooc_qt_picklefilename} does not exist, constructing quadtree from scratch')
```

davemfish replied:
Done.

davemfish commented on src/natcap/invest/recreation/recmodel_server.py:
```diff
@@ -148,104 +339,140 @@ def get_valid_year_range(self):
+                     if not file.endswith('.zip')]
+        if len(file_list) > 1:
+            # if it's a multifile AOI, assume a shapefile
+            aoi_path = glob.glob(os.path.join(workspace_path, '*.shp'))[0]
```
Hmm, that's interesting. I think that GPKG `wal` file situation is pretty unlikely because the AOI that get's zipped and sent here is actually a copy of the user-input one. So the user would have to start the model and open one of the intermediate files at just the right time.

But now that you mention it, I think I subsequently changed the "aoi prep" and "aoi grid" tasks to always create a GPKG. So I think this whole thing can be simplified to just find the file that `endswith(".gpkg")`.

davemfish replied:
On second thought, I'm just going to pass the filename from the client.

emlys replied:
Looks like maybe this hasn't been updated yet?

emlys commented on scripts/recreation_server/setup_vm.sh:
```diff
@@ -0,0 +1,52 @@
+
+# After setting up a new VM or new server cache,
+# initialize a cron job that periodically clears out the cache.
+# Copy the script so that updates to the inveset repo don't clobber it
```
```suggestion
# Copy the script so that updates to the invest repo don't clobber it
```

davemfish commented on tests/test_recreation.py:
```diff
@@ -584,48 +444,418 @@ def test_numpy_pickling_queue(self):
+            numpy.testing.assert_almost_equal(
+                actual_scenario_sums[key], expected_scenario_sums[key], decimal=3)
+
+        # TODO: assert that all tabular outputs are indexed by the same poly_id
```
I never ended up creating a `poly_id`, I just relied on the FIDs, but I think maybe it is a good idea to come back and do this. Based on feedback from Claire too: https://github.com/natcap/invest/pull/1832#discussion_r2004405523

davemfish commented on tests/test_recreation.py:
```diff
@@ -11,8 +11,12 @@
 import functools
```
@claire-simpson , my opinion is that `_assemble_regression_data` and `_retrieve_user_days`, and nearly all of those functions listed are exercised fairly well by existing tests. `test_all_metrics_local_server` asserts on the `regression_data.gpkg`, which is the main output of the `_assemble` function. And it also asserts on the userdays results which are from `_retrieve_user_days`.

I do think there can be benefit of having unit tests, but in this case I don't feel like the cost of having more tests to maintain gets us too much added value. If I were writing this model from scratch and building these functions, I would likely write unit tests for them as I go, to make sure each function does what's intended. And I think I took that approach here with some of the new functions that were added in this PR. Since we're already starting from a functional model, and have "regression" tests that run the whole thing from start to finish, I feel okay about not going back to write unit tests for those existing functions.

I'm happy to reconsider though if there are some specific cases/conditions that you think would be useful to test that we're not testing yet.

claire-simpson replied:
This makes sense. I was assuming that adding tests for these 2 functions would make it easier for us but wasn't really considering the work that maintaining tests requires. I'm good with leaving the tests as is

claire-simpson commented on src/natcap/invest/recreation/recmodel_client.py:
```diff
@@ -1567,15 +1572,20 @@ def _build_regression(
+    if (response_array == 0).any():
+        epsilon = response_array[response_array > 0].min() / 2
+        response_array[response_array == 0] = epsilon
+    transformed_array = numpy.log(response_array / (1 - response_array))
```
Nice, I like this way of adjusting only the 0's

claire-simpson reviewed (approved):
Thanks @davemfish, this looks good!

emlys reviewed (changes requested):
This is looking good @davemfish! I just had a few more comments. They might be kind of buried among all the comment threads, but I think I have marked everything else as resolved

davemfish commented:
> This is looking good @davemfish! I just had a few more comments. They might be kind of buried among all the comment threads, but I think I have marked everything else as resolved

Thanks! I added a comment to a shell script, updated a logger message in `recmodel_server`, and updated the Makefile.

emlys commented on src/natcap/invest/recreation/recmodel_server.py:
```diff
@@ -148,203 +344,214 @@ def get_valid_year_range(self):
+
         aoi_layer = aoi_vector.GetLayer()
-        aoi_extent = aoi_layer.GetExtent()
+        # aoi_extent = aoi_layer.GetExtent()
```
Unneeded line?

emlys commented:
@davemfish, sorry, I had one more comment that maybe github is hiding? Mainly I think it would be good for the server to validate that the requested dates are valid. 
https://github.com/natcap/invest/pull/1832#discussion_r2008273372

I noticed one more commented line that can probably be removed. Also, it occurred to me that the API and code might read a little better if the `date_range` tuple was replaced with separate `start_date` and `end_date` args in all the functions that use it. But this is minor feedback coming pretty late, so feel free to disregard.

emlys reviewed (approved):
Thanks @davemfish! Looks great
